api/store.proto

```diff
@@ -0,0 +1,185 @@
+	rpc GetObject(GetObjectRequest) returns (Object) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// FindObjects returns a set of objects of a given type, according to
+	// selectors included in the request.
+	rpc FindObjects(FindObjectsRequest) returns (FindObjectsResponse) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// CreateObject creates the specified object. The ID and Meta fields
+	// should be left empty, and will be filled in and included in the
+	// returned object.
+	rpc CreateObject(Object) returns (Object) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// UpdateObjects atomically commits the supplied objects to the data store.
+	// - Returns `FailedPrecondition` and does not commit any changes if
+	//   any of the objects have changed since they were retrieved.
+	rpc UpdateObjects(UpdateObjectsRequest) returns (UpdateObjectsResponse) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// Watch starts a stream that returns any changes to objects that match
+	// the specified selectors. When the stream begins, it immediately sends
+	// an empty message back to the client. It is important to wait for
+	// this message before taking any actions that depend on an established
+	// stream of changes for consistency.
+	rpc Watch(WatchRequest) returns (stream WatchMessage) {
+		option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	};
+
+	// View creates a snapshot of the current state of the store, and
+	// accepts a stream of queries against it. Each query returns a
+	// response in the response stream.
+	// TODO(aaronl): Does it make sense to support this in the API?
+	//rpc View(stream ViewRequest) returns (stream ViewResponse) {
+	//	option (docker.protobuf.plugin.tls_authorization) = { roles: "swarm-manager" };
+	//}
+}
+
+message GetObjectRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// ID of the object to retrieve.
+	string object_id = 2 [(gogoproto.customname) = "ObjectID"];
+}
+
+message FindObjectsRequest {
+	// Kind can contain a builtin type such as "node", "secret", etc. or
+	// the kind specified by a custom-defined object.
+	string kind = 1;
+
+	// Selectors are combined with OR logic. An empty list of selectors
+	// causes all objects of the given type to be returned.
+	repeated SelectBy selectors = 2;
+}
+
+message FindObjectsResponse {
+	repeated Object results = 1;
+}
+
+message UpdateObjectsRequest {
+	repeated Object objects = 1;
+}
+
+message UpdateObjectsResponse {
+	// Objects contains the updated objects. On a successful return, these
+	// will be the same as the objects passed in UpdateObjectsRequest,
+	// except with Meta updated.
+	repeated Object objects = 1;
```

Comment:
I think it's useful to get the updated object back because you may need to know the new version. If you want to update it again, you'll need to know that version, otherwise the sequencer would reject the update.

I suppose we could just return the versions, but it might be confusing to figure out which version corresponds to which object. It seemed simpler to just return the full objects.